pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillGroupByDataTableReducer.java

```diff
@@ -0,0 +1,725 @@
+        for (int i = 0; i < limit && sortedIterator.hasNext(); i++) {
+          Object[] row = sortedIterator.next().getValues();
+          extractFinalAggregationResults(row);
+          for (int j = 0; j < numColumns; j++) {
+            row[j] = columnDataTypes[j].convert(row[j]);
+          }
+          rows.add(row);
+        }
+      }
+      DataSchema resultDataSchema = postAggregationHandler.getResultDataSchema();
+      ColumnDataType[] resultColumnDataTypes = resultDataSchema.getColumnDataTypes();
+      int numResultColumns = resultColumnDataTypes.length;
+      int numResultRows = rows.size();
+      List<Object[]> resultRows = new ArrayList<>(numResultRows);
+      for (Object[] row : rows) {
+        Object[] resultRow = postAggregationHandler.getResult(row);
+        for (int i = 0; i < numResultColumns; i++) {
+          resultRow[i] = resultColumnDataTypes[i].format(resultRow[i]);
+        }
+        resultRows.add(resultRow);
+        _primaryKeys.add(constructKey(resultRow));
+      }
+      List<Object[]> gapfillResultRows = gapFill(resultRows, resultColumnDataTypes);
+      brokerResponseNative.setResultTable(new ResultTable(resultDataSchema, gapfillResultRows));
+    } else {
+      // PQL query with SQL group-by mode and response format
+      // NOTE: For PQL query, keep the order of columns as is (group-by expressions followed by aggregations), no need
+      //       to perform post-aggregation or filtering.
+
+      for (int i = 0; i < limit && sortedIterator.hasNext(); i++) {
+        Object[] row = sortedIterator.next().getValues();
+        extractFinalAggregationResults(row);
+        for (int j = 0; j < numColumns; j++) {
+          row[j] = columnDataTypes[j].convertAndFormat(row[j]);
+        }
+        rows.add(row);
+      }
+      brokerResponseNative.setResultTable(new ResultTable(prePostAggregationDataSchema, rows));
+    }
+  }
+
+  List<Object[]> gapFill(List<Object[]> resultRows, ColumnDataType[] resultColumnDataTypes) {
+    int limit = _queryContext.getLimit();
+    int numResultColumns = resultColumnDataTypes.length;
+    List<Object[]> gapfillResultRows = new ArrayList<>(limit);
+    long step = _dateTimeGranularity.granularityToMillis();
+    int index = 0;
+    for (long time = _startMs; time + 2 * step <= _endMs; time += step) {
+      Set<Key> keys = new HashSet<>(_primaryKeys);
+      while (index < resultRows.size()) {
+        long timeCol = _dateTimeFormatter.fromFormatToMillis((String) resultRows.get(index)[0]);
+        if (timeCol < time) {
+          index++;
+        } else if (timeCol == time) {
+          gapfillResultRows.add(resultRows.get(index));
+          if (gapfillResultRows.size() == limit) {
+            return gapfillResultRows;
+          }
+          Key key = constructKey(resultRows.get(index));
+          keys.remove(key);
+          for (int colIndex = 2; colIndex < numResultColumns; colIndex++) {
+            _previous.put(key, resultRows.get(index));
+          }
```

Comment:
this loop performs the same operation in each iteration. Should this be `resultRows.get(colIndex)`? Are these lines covered by tests?

Reply:
This is the bug. We just need put resultRows.get(colIndex) inside the _previous once.